File: src/app/services/leaverequest.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';

export interface LeaveRequest {
    id: string;
    driverId: string;
    leaveTypeEnum: LeaveType;
    leaveTypeName: string;
    startDate: string;
    endDate: string;
    totalDays: number;
    reason: string;
    status: LeaveStatus;
    approvedBy?: string;
    approvedAt?: string;
    rejectionReason?: string;
    createdAt: string;
    updatedAt: string;
    driver?: {
        id: string;
        name: string;
        email: string;
    };
    approverUser?: {
        id: string;
        name: string;
        email: string;
    };
}

export enum LeaveType {
    Annual = 1,
    Sick = 2,
    Personal = 3,
    Emergency = 4,
    Maternity = 5,
    Paternity = 6,
    Bereavement = 7,
    Other = 8
}

export enum LeaveStatus {
    Pending = 0,
    Approved = 1,
    Rejected = 2,
    Cancelled = 3
}

export interface CreateLeaveRequestDto {
    driverId: string;
    leaveType: LeaveType;
    startDate: string;
    endDate: string;
    reason: string;
}

export interface LeaveBalance {
    annual: LeaveBalanceItem;
    sick: LeaveBalanceItem;
    personal: LeaveBalanceItem;
}

export interface LeaveBalanceItem {
    used: number;
    total: number;
    remaining: number;
}

export interface ApproveLeaveRequestDto {
    approvedBy: string;
    notes?: string;
}

export interface RejectLeaveRequestDto {
    rejectedBy: string;
    rejectionReason: string;
}

export interface LeaveTypeInfo {
    name: string;
    description: string;
}

@Injectable({
    providedIn: 'root'
})
export class LeaveRequestService {
    private readonly apiUrl = 'http://localhost:5129/api/leaverequests';

    private httpOptions = {
        headers: new HttpHeaders({
            'Content-Type': 'application/json'
        })
    };

    constructor(private http: HttpClient) { }

    getAllLeaveRequests(): Observable<LeaveRequest[]> {
        return this.http.get<LeaveRequest[]>(this.apiUrl).pipe(catchError(this.handleError));
    }

    getLeaveRequestsByDriverId(driverId: string): Observable<LeaveRequest[]> {
        return this.http.get<LeaveRequest[]>(`${this.apiUrl}/driver/${driverId}`).pipe(catchError(this.handleError));
    }

    getLeaveRequestById(id: string): Observable<LeaveRequest> {
        return this.http.get<LeaveRequest>(`${this.apiUrl}/${id}`).pipe(catchError(this.handleError));
    }

    createLeaveRequest(dto: CreateLeaveRequestDto): Observable<LeaveRequest> {
        return this.http.post<LeaveRequest>(this.apiUrl, dto, this.httpOptions).pipe(catchError(this.handleError));
    }

    approveLeaveRequest(id: string, dto: ApproveLeaveRequestDto): Observable<LeaveRequest> {
        return this.http.put<LeaveRequest>(`${this.apiUrl}/${id}/approve`, dto, this.httpOptions).pipe(catchError(this.handleError));
    }

    rejectLeaveRequest(id: string, dto: RejectLeaveRequestDto): Observable<LeaveRequest> {
        return this.http.put<LeaveRequest>(`${this.apiUrl}/${id}/reject`, dto, this.httpOptions).pipe(catchError(this.handleError));
    }

    deleteLeaveRequest(id: string): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`).pipe(catchError(this.handleError));
    }

    getLeaveBalance(driverId: string): Observable<LeaveBalance> {
        return this.http.get<LeaveBalance>(`${this.apiUrl}/balance/${driverId}`).pipe(catchError(this.handleError));
    }

    getLeaveTypes(): Observable<LeaveTypeInfo[]> {
        return this.http.get<LeaveTypeInfo[]>(`${this.apiUrl}/types`).pipe(catchError(this.handleError));
    }

    getLeaveTypeString(leaveType: LeaveType): string {
        return LeaveType[leaveType];
    }

    getLeaveStatusString(status: LeaveStatus): string {
        return LeaveStatus[status];
    }

    getStatusClass(status: LeaveStatus): string {
        switch (status) {
            case LeaveStatus.Approved: return 'status-approved';
            case LeaveStatus.Pending: return 'status-pending';
            case LeaveStatus.Rejected: return 'status-rejected';
            case LeaveStatus.Cancelled: return 'status-cancelled';
            default: return '';
        }
    }

    getTypeClass(leaveType: LeaveType): string {
        switch (leaveType) {
            case LeaveType.Annual: return 'type-annual';
            case LeaveType.Sick: return 'type-sick';
            case LeaveType.Personal: return 'type-personal';
            case LeaveType.Emergency: return 'type-emergency';
            case LeaveType.Maternity:
            case LeaveType.Paternity: return 'type-family';
            case LeaveType.Bereavement: return 'type-bereavement';
            default: return 'type-other';
        }
    }

    calculateBusinessDays(startDate: Date, endDate: Date): number {
        let count = 0;
        const current = new Date(startDate);
        while (current <= endDate) {
            const dayOfWeek = current.getDay();
            if (dayOfWeek !== 0 && dayOfWeek !== 6) count++;
            current.setDate(current.getDate() + 1);
        }
        return count;
    }

    validateLeaveRequest(dto: CreateLeaveRequestDto): string[] {
        const errors: string[] = [];
        const startDate = new Date(dto.startDate);
        const endDate = new Date(dto.endDate);
        const today = new Date();
        today.setHours(0, 0, 0, 0);

        if (startDate < today) errors.push('Start date cannot be in the past');
        if (endDate < startDate) errors.push('End date cannot be before start date');

        if (dto.leaveType !== LeaveType.Emergency) {
            const advanceNotice = new Date();
            advanceNotice.setDate(today.getDate() + 1);
            if (startDate < advanceNotice) errors.push('Leave requests require at least 1 day advance notice');
        }

        if (!dto.reason || dto.reason.trim().length < 10) errors.push('Please provide a detailed reason (minimum 10 characters)');
        return errors;
    }

    private handleError(error: any): Observable<never> {
        console.error('An error occurred:', error);
        let errorMessage = 'An unexpected error occurred';
        if (error.error?.message) errorMessage = error.error.message;
        else if (error.message) errorMessage = error.message;
        return throwError(() => new Error(errorMessage));
    }
}

File: FleetManagerPro.API/Services/LeaveRequestService.cs
using FleetManagerPro.API.Models;
using FleetManagerPro.API.Data.Repository;
using FleetManagerPro.API.DTOs;

namespace FleetManagerPro.API.Services
{
    public interface ILeaveRequestService
    {
        Task<List<LeaveRequest>> GetAllLeaveRequestsAsync();
        Task<List<LeaveRequest>> GetLeaveRequestsByDriverIdAsync(string driverId);
        Task<LeaveRequest?> GetLeaveRequestByIdAsync(string id);
        Task<LeaveRequest> CreateLeaveRequestAsync(CreateLeaveRequestDto dto);
        Task<LeaveRequest?> ApproveLeaveRequestAsync(string id, string approvedBy, string? notes = null);
        Task<LeaveRequest?> RejectLeaveRequestAsync(string id, string rejectedBy, string rejectionReason);
        Task<bool> DeleteLeaveRequestAsync(string id);
        Task<LeaveBalanceDto> GetLeaveBalanceAsync(string driverId);
        Task<List<LeaveTypeDto>> GetLeaveTypesAsync();
    }

    public class LeaveRequestService : ILeaveRequestService
    {
        private readonly ILeaveRequestRepository _leaveRequestRepository;
        private readonly ILogger<LeaveRequestService> _logger;

        public LeaveRequestService(ILeaveRequestRepository leaveRequestRepository, ILogger<LeaveRequestService> logger)
        {
            _leaveRequestRepository = leaveRequestRepository;
            _logger = logger;
        }

        public async Task<List<LeaveRequest>> GetAllLeaveRequestsAsync()
        {
            return await _leaveRequestRepository.GetAllAsync();
        }

        public async Task<List<LeaveRequest>> GetLeaveRequestsByDriverIdAsync(string driverId)
        {
            return await _leaveRequestRepository.GetByDriverIdAsync(driverId);
        }

        public async Task<LeaveRequest?> GetLeaveRequestByIdAsync(string id)
        {
            return await _leaveRequestRepository.GetByIdAsync(id);
        }

        public async Task<LeaveRequest> CreateLeaveRequestAsync(CreateLeaveRequestDto dto)
        {
            var totalDays = (dto.EndDate.Date - dto.StartDate.Date).Days + 1;

            var leaveBalance = await GetLeaveBalanceAsync(dto.DriverId);
            var currentTypeBalance = dto.LeaveType switch
            {
                Models.LeaveType.Annual => leaveBalance.Annual,
                Models.LeaveType.Sick => leaveBalance.Sick,
                Models.LeaveType.Personal => leaveBalance.Personal,
                _ => new LeaveBalanceItem { Used = 0, Total = 999, Remaining = 999 }
            };

            if (totalDays > currentTypeBalance.Remaining && dto.LeaveType != Models.LeaveType.Emergency)
            {
                throw new InvalidOperationException($"Insufficient {dto.LeaveType} leave balance. Available: {currentTypeBalance.Remaining} days");
            }

            var leaveRequest = new LeaveRequest
            {
                Id = Guid.NewGuid().ToString(),
                DriverId = dto.DriverId,
                LeaveTypeEnum = dto.LeaveType,
                StartDate = dto.StartDate,
                EndDate = dto.EndDate,
                TotalDays = totalDays,
                Reason = dto.Reason,
                StatusEnum = LeaveStatus.Pending,
                SubmittedDate = DateTime.UtcNow,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            return await _leaveRequestRepository.CreateAsync(leaveRequest);
        }

        public async Task<LeaveRequest?> ApproveLeaveRequestAsync(string id, string approvedBy, string? notes = null)
        {
            var leaveRequest = await _leaveRequestRepository.GetByIdAsync(id);
            if (leaveRequest == null) return null;

            if (leaveRequest.StatusEnum != LeaveStatus.Pending)
            {
                throw new InvalidOperationException("Only pending leave requests can be approved");
            }

            leaveRequest.StatusEnum = LeaveStatus.Approved;
            leaveRequest.ApprovedBy = approvedBy;
            leaveRequest.ApprovedDate = DateTime.UtcNow;
            leaveRequest.UpdatedAt = DateTime.UtcNow;

            if (!string.IsNullOrEmpty(notes))
            {
                leaveRequest.RejectionReason = notes;
            }

            return await _leaveRequestRepository.UpdateAsync(leaveRequest);
        }

        public async Task<LeaveRequest?> RejectLeaveRequestAsync(string id, string rejectedBy, string rejectionReason)
        {
            var leaveRequest = await _leaveRequestRepository.GetByIdAsync(id);
            if (leaveRequest == null) return null;

            if (leaveRequest.StatusEnum != LeaveStatus.Pending)
            {
                throw new InvalidOperationException("Only pending leave requests can be rejected");
            }

            leaveRequest.StatusEnum = LeaveStatus.Rejected;
            leaveRequest.ApprovedBy = rejectedBy;
            leaveRequest.ApprovedDate = DateTime.UtcNow;
            leaveRequest.RejectionReason = rejectionReason;
            leaveRequest.UpdatedAt = DateTime.UtcNow;

            return await _leaveRequestRepository.UpdateAsync(leaveRequest);
        }

        public async Task<bool> DeleteLeaveRequestAsync(string id)
        {
            var leaveRequest = await _leaveRequestRepository.GetByIdAsync(id);
            if (leaveRequest == null) return false;

            if (leaveRequest.StatusEnum != LeaveStatus.Pending)
            {
                throw new InvalidOperationException("Only pending leave requests can be deleted");
            }

            return await _leaveRequestRepository.DeleteAsync(id);
        }

        public async Task<LeaveBalanceDto> GetLeaveBalanceAsync(string driverId)
        {
            var currentYear = DateTime.UtcNow.Year;

            var approvedRequests = await _leaveRequestRepository.GetApprovedByDriverIdAndYearAsync(driverId, currentYear);

            var usedAnnual = approvedRequests.Where(lr => lr.LeaveTypeEnum == Models.LeaveType.Annual).Sum(lr => lr.TotalDays);
            var usedSick = approvedRequests.Where(lr => lr.LeaveTypeEnum == Models.LeaveType.Sick).Sum(lr => lr.TotalDays);
            var usedPersonal = approvedRequests.Where(lr => lr.LeaveTypeEnum == Models.LeaveType.Personal).Sum(lr => lr.TotalDays);

            const int annualAllowance = 21;
            const int sickAllowance = 10;
            const int personalAllowance = 5;

            return new LeaveBalanceDto
            {
                Annual = new LeaveBalanceItem
                {
                    Used = usedAnnual,
                    Total = annualAllowance,
                    Remaining = annualAllowance - usedAnnual
                },
                Sick = new LeaveBalanceItem
                {
                    Used = usedSick,
                    Total = sickAllowance,
                    Remaining = sickAllowance - usedSick
                },
                Personal = new LeaveBalanceItem
                {
                    Used = usedPersonal,
                    Total = personalAllowance,
                    Remaining = personalAllowance - usedPersonal
                }
            };
        }

        public async Task<List<LeaveTypeDto>> GetLeaveTypesAsync()
        {
            return await Task.FromResult(new List<LeaveTypeDto>
            {
                new LeaveTypeDto { Name = "Annual", Description = "Annual vacation leave" },
                new LeaveTypeDto { Name = "Sick", Description = "Sick leave" },
                new LeaveTypeDto { Name = "Personal", Description = "Personal leave" },
                new LeaveTypeDto { Name = "Emergency", Description = "Emergency leave" },
                new LeaveTypeDto { Name = "Maternity", Description = "Maternity leave" },
                new LeaveTypeDto { Name = "Paternity", Description = "Paternity leave" },
                new LeaveTypeDto { Name = "Bereavement", Description = "Bereavement leave" }
            });
        }
    }
}